### DISCUSSION TITLE
Allow for procedural methods and fields in C#

### DISCUSSION THREAD

**Discussion by LuckyScamper** [dotnet/csharplang#6988] [Language Ideas]
**Note:** technically this is already done a bit since [Top-Level Statements](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/program-structure/top-level-statements) were introduced.

**Goal:** allow users to use fields, methods or extension methods outside of classes in C#, e.g. what C++/C/JS and other languages have. 

**Why?:** [OOP has significant limitations that introduce needless complexities, especially in terms of encapsulation, code verbosity and so on.](https://www.youtube.com/watch?v=QM1iUe6IofM&ab_channel=BrianWill) It is purely so more design patterns can be introduced within a project, and to prevent having to write left-hand assignments of the static container over and over.

**What does this mean for the language and compiler?**: although it would go against the core principle of C# being a pure OOP language, it wouldn't change all that much. As far as the compiler is concerned, fields and methods would simply be put in a hidden container (partial static class) based on their current namespace (including the global namespace). They would therefore still be a normal managed resources as any other. Naming collisions wouldn't be an issue any more than they are today if you were to use libraries with the same namespace and the same class names. CS0101 would therefore also apply to methods/fields.

**Potential pitfalls:** if such a feature were to be accepted, it would be obvious people would want to use them for .NET Standard Libraries, which in turn would require a new version of .NET Standard which IIRC is out of the question.

**Code examples:**

in the following example, (copied from .net source code) we will calculate the angle between two points.

      
      namespace Foo 
      {
            public static class FooMath
            {
                  public static double AngleBetween(Vector vector1, Vector vector2)
                  {
                      double sin = vector1._x * vector2._y - vector2._x * vector1._y;  
                      double cos = vector1._x * vector2._x + vector1._y * vector2._y;
                  
                      return Math.Atan2(sin, cos) * (180 / Math.PI);
                  }
            }
      }

In its usage, one would have to write:

        using Foo;
        ...
        var angle = FooMath.AngleBetween(bar1, bar2);

For this new feature, the code could be simplified as follows:

            public double AngleBetween(Vector vector1, Vector vector2)
            {
                double sin = vector1._x * vector2._y - vector2._x * vector1._y;  
                double cos = vector1._x * vector2._x + vector1._y * vector2._y;
            
                return Math.Atan2(sin, cos) * (180 / Math.PI);
            }

The compilers sees that the method is outside of a class, and therefore assumes it (correctly) to be _static_. Note how Math is still a static class; the use for procedural methods/fields is purely optional, and for readability and organization there is no need to not have them in a shorthand 'Math' container. Doing otherwise would also break earlier versions of C# and .NET for no good reason. If the user still declares the method static the compiler will swallow it, but tools like Resharper would call the keyword there redundant. (in a similar vein to IDE0001 or IDE0005 warnings)

The method, because it isn't placed in a static class by the user, can also not be private or protected, but only use public or internal access modifiers. (or file, if that is applied.)

In it usage, the code would be shortened to:

        using Foo;
        ...
        var angle = AngleBetween(bar1, bar2);

Extension methods, generics and unsafes can be done the same way:

            public double Bar(this Foo foo)
            {
                return foo.Foo1 * 2;
            }

            public T CreateInstance<T>(string email) where T : IHasEmail, new()
            {
                  var item = new T();
                  item.Email  = email;
                  return item;
            }

            unsafe float FastInverseSquareRoot(float x)
              {
                  float xhalf = 0.5f * x;
                  int i = *(int*)&x;
                  i = 0x5f3759df - (i >> 1);
                  x = *(float*)&i;
                  x = x * (1.5f - xhalf * x * x);
                  return x;
              }

**Conclusion**: It is a feature designed to remove scaffolding (especially left-hand calling of the static class container for a method/field) and code verbosity, and allow for easier use of other design patterns outside of pure OOP programming. In terms of optimization for the compiler this would result in a very slight improvement. This is as all methods/fields inside of a namespace within a project (library, console, web, etc) would be placed in a single shared static class, rather than the usual multiple static classes users create.

**Comment by TahirAhmadov** [comment]
You can already do this:
```cs
class Procedures
{
  public void Foo(string s, int x) { .. }
}

// in your GlobalSuppressions.cs file (or something like that)
global using static Procedures;

// then elsewhere
Foo("abc", 123); // no problem
```

**Reply by LuckyScamper** [reply]
That would just be more boilerplate

**Comment by HaloFour** [comment]
The C# team has expressed that they're [not interested](https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-09-28.md#top-level-functions) in exploring top-level members beyond what they did for "simple programs".  Yes, putting the members into a static class is boilerplate, but not considered onerous boilerplate.

**Comment by siegfriedpammer** [comment]
> although it would go against the core principle of C# being a pure OOP language

C# isn't (and never was) a pure OOP language. There's always been the functional aspect starting with delegates in version 1.

**Reply by 333fred** [reply]
Did you mean isn't?

**Reply by siegfriedpammer** [reply]
Yes
